doc/architecture/decisions/0015-handle-nomis-users-without-inheritance-in-cas2.md

```diff
@@ -0,0 +1,201 @@
+
+```json
+{
+  "username": "testuser1",
+  "staffId": 324323,
+  "firstName": "John",
+  "lastName": "Smith",
+  "activeCaseloadId": "BXI",
+  "accountStatus": "OPEN",
+  "accountType": "GENERAL",
+  "primaryEmail": "test@test.com",
+  "dpsRoleCodes": ["string"],
+  "accountNonLocked": true,
+  "credentialsNonExpired": true,
+  "enabled": true,
+  "admin": true,
+  "active": true,
+  "staffStatus": "ACTIVE"
+}
+```
+
+and others are very much specific to Delius.
+
+The CAS API makes use of [JPA Inheritance](https://thorben-janssen.com/complete-guide-inheritance-strategies-jpa-hibernate/) 
+to handle CAS1/2/3 specific flavours of "Application". An `Application` belongs 
+to a `User` (a Delius user).
+
+In handling Nomis users we have two polymorphism/inheritance challenges to 
+either tackle or sidestep:
+
+1. Allow `Application`s to be associated with either Delius or Nomis users.
+2. Allow authorisation code to handle both Nomis and Delius users.
+
+NB: some classes such as `UserAccessService` would need to handle 
+both the polymorphic `Application` and a new polymorphic `User`.
+
+We have considered two approaches:
+
+### 1. **Introduce more JPA inheritance**
+
+In this option we would associate users with an abstract class 
+such as `AuthenticatedUser`s rather than Delius `User`s.
+
+This would require changes to existing classes and concepts to allow them 
+to cope with:
+
+- `Application`s associated with either Delius or Nomis users
+
+- `UserDetail` representations coming from either Delius or Nomis
+
+- Authorisation/access checks (LAO) coming from either Delius or Nomis 
+  in an `UserAccessService` class or similar
+
+#### Advantages
+- retains existing pattern for handling service-specific needs
+
+#### Disadvantages
+- increases complexity and size of classes and functions which are already 
+  over-long
+
+- requires extensive refactoring of code used by existing live CAS 
+  services (1 and 3). Despite good test coverage this is risky, due 
+  to the need to migrate existing application and user records to the new formats
+
+### 2. **Introduce CAS2 specific classes**
+
+In this (preferred) option we would handle Nomis users and their applications with CAS2-specific classes, e.g.
+
+- `Cas2ApplicationEntity`
+- `Cas2JsonSchemaService`
+- `NomisUserDetail`
+- `NomisUserEntity`
+- `NomisUserService`
+- `NomisUserAccessService`
+
+#### Advantages
+
+- is a low tech approach with does not require sophisticated use of JPA 
+  inheritance. It's therefore easier for a broad range of developers to 
+  understand and maintain
+
+- results in reduction in size and complexity of large classes and 
+  functions (they will be left dealing "only" with CAS1 and CAS3). In 
+  this way it drives a desirable refactoring of the codebase to leave 
+  it in a clearer state with the CAS2 service-specific code largely 
+  decoupled into discrete classes.
+
+#### Disadvantages
+
+- there will be more classes and repeated code
```

Comment:
thks Tom: `UserAccessService` is an interesting case -- though from where I'm sitting perhaps not an example of something which can "be written once and apply to all users"!

Take:

```kotlin
  fun userCanViewApplication(user: UserEntity, application: ApplicationEntity): Boolean {
    if (user.id == application.createdByUser.id) {
      return true
    }

    return when (application) {
      is ApprovedPremisesApplicationEntity -> userCanViewApprovedPremisesApplicationCreatedBySomeoneElse(user, application)
      is TemporaryAccommodationApplicationEntity -> userCanViewTemporaryAccommodationApplicationCreatedBySomeoneElse(user, application)
      else -> false
    }
```

This is currently typed to both Delius users and the polymorphic applications. My initial plan with this will be to implement a `NomisUserAccessService` which has:

```kotlin
fun userCanViewCas2Application(user: NomisUserEntity, application: Cas2ApplicationEntity): Boolean` {
  // CAS2 auth rules
  // AND does the LAO test against PrisonAPI rather than Community API using the NomisUserEntity
}
```
When the day comes that CAS3 needs to grant access to Nomis users you could then add:

```kotlin
fun userCanViewCas3Application(user: NomisUserEntity, application: Cas3ApplicationEntity): Boolean {
  // CAS3 auth rules
 // AND the LAO test to Prison API with NomisUserEntiry
}
```

Does that make sense? But perhaps I'm getting the wrong end of the stick and getting stuck on the fact that LAO is one of the features which is entirely different for Nomis versus Delius users?